A variety of Cppcheck fixes

jclarkeSTFC opened the pull request:
Several Cppcheck fixes.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Delete unnecessary virtual destructors.

jclarkeSTFC committed:
Fixes for ImportMDHistoWorkspaceBase

The member variable nbins is unused, and was shadowed inside a method.
Also made some variables in that method const.

jclarkeSTFC committed:
AlgorithmDialog missing const in a few places

jclarkeSTFC committed:
Fix some return by reference in FitPropertyBrowser

jclarkeSTFC committed:
Fix several passedByValue Cppcheck warnings

jclarkeSTFC committed:
Further Cppcheck fixes

thomashampson reviewed (changes requested):
Looks fairly straightforward, lots of small improvements by returning large objects by reference. Just one place where it looks like `const` is missing.

jclarkeSTFC committed:
Argument may be modified so pass by value

Passing-by-value is the original behaviour, and `lower_bound` does not
specify `const`, hence we can't use a `const` reference here.

jclarkeSTFC commented on Framework/Algorithms/inc/MantidAlgorithms/ExtractMaskToTable.h:
```diff
@@ -35,7 +35,7 @@ class MANTID_ALGORITHMS_DLL ExtractMaskToTable final : public API::Algorithm {
 
   /// Remove the items appeared in a vector from another
-  std::vector<detid_t> subtractVector(std::vector<detid_t> minuend, std::vector<detid_t> subtrahend);
+  std::vector<detid_t> subtractVector(const std::vector<detid_t> &minuend, std::vector<detid_t> &subtrahend);
```
Yes, thanks for spotting that, I've changed `subtrahend` back to pass-by-value because `lower_bound` does not specify `const` for its arguments.
